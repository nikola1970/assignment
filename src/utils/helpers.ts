export function normalizeData<T extends { id: string }>(arr: Array<T>): { [key: string]: T } {
    return arr.reduce((obj: { [key: string]: T }, item: T) => {
        obj[item.id] = item;
        return obj;
    }, {});
}

export function isObjectEmpty(obj: { [key: string]: unknown }) {
    return Object.keys(obj).length === 0;
}

export function formatUsers(userCount: string | number) {
    return `${userCount} Users`;
}

export function formatTimezone(date: string) {
    return new Date(date).toLocaleString();
}

export function formatViews(views: number | string) {
    return `${views} ${views === 1 ? 'View' : 'Views'}`;
}
