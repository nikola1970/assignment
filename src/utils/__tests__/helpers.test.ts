import { normalizeData, isObjectEmpty, formatUsers, formatTimezone, formatViews } from '../helpers';

// mockData
import { locations } from '../../testing/msw/mockData';

describe('helpers tests', () => {
    it('should normalize data', () => {
        expect(normalizeData(locations)).toMatchSnapshot();
    });

    it('should check for empty objects', () => {
        expect(isObjectEmpty({})).toBe(true);
        expect(isObjectEmpty({ test: 'test' })).toBe(false);
    });

    it('should properly format user count', () => {
        const expected = '653 Users';
        expect(formatUsers(653)).toBe(expected);
    });

    it('should properly format time', () => {
        const expected = '05/03/2021, 10:10:35';
        expect(formatTimezone(locations[0].createdAt)).toBe(expected);
    });

    it('should properly format views', () => {
        // only one
        const expected = '1 View';
        expect(formatViews(1)).toBe(expected);

        // zero or multiple
        const expected2 = '2 Views';
        expect(formatViews(2)).toBe(expected2);
    });
});
