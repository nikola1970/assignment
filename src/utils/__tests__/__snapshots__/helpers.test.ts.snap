// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers tests should normalize data 1`] = `
Object {
  "1": Object {
    "createdAt": "2021-03-05T09:10:35.462Z",
    "description": "Investment Account",
    "id": "1",
    "name": "McGlynnview",
    "userCount": 6221,
    "views": 0,
  },
  "10": Object {
    "createdAt": "2021-03-04T12:28:55.319Z",
    "description": "Rustic",
    "id": "10",
    "name": "South Freidabury",
    "userCount": 10435,
    "views": 0,
  },
  "11": Object {
    "createdAt": "2021-03-05T07:53:36.790Z",
    "description": "Mobility Auto Loan Account Chief",
    "id": "11",
    "name": "Turnerland",
    "userCount": 22083,
    "views": 0,
  },
  "12": Object {
    "createdAt": "2021-03-04T18:48:49.125Z",
    "description": "Intelligent Bike",
    "id": "12",
    "name": "New Dewayneview",
    "userCount": 83058,
    "views": 0,
  },
  "13": Object {
    "createdAt": "2021-03-04T14:09:13.414Z",
    "description": "withdrawal Money Market Account",
    "id": "13",
    "name": "New Nicolasberg",
    "userCount": 8570,
    "views": 0,
  },
  "14": Object {
    "createdAt": "2021-03-05T03:33:37.286Z",
    "description": "Cheese Industrial",
    "id": "14",
    "name": "Sydneeland",
    "userCount": 2136,
    "views": 0,
  },
  "15": Object {
    "createdAt": "2021-03-05T02:17:35.519Z",
    "description": "Beauty",
    "id": "15",
    "name": "Carterbury",
    "userCount": 1198,
    "views": 0,
  },
  "16": Object {
    "createdAt": "2021-03-04T15:18:50.194Z",
    "description": "Frozen integrate Communications",
    "id": "16",
    "name": "North Axel",
    "userCount": 34091,
    "views": 0,
  },
  "17": Object {
    "createdAt": "2021-03-05T05:32:43.805Z",
    "description": "Towels",
    "id": "17",
    "name": "Abernathystad",
    "userCount": 54825,
    "views": 0,
  },
  "18": Object {
    "createdAt": "2021-03-05T06:38:34.971Z",
    "description": "magenta override transmitting",
    "id": "18",
    "name": "East Dahlia",
    "userCount": 33423,
    "views": 0,
  },
  "19": Object {
    "createdAt": "2021-03-05T06:56:03.030Z",
    "description": "Right-sized AI",
    "id": "19",
    "name": "Crystalville",
    "userCount": 14558,
    "views": 0,
  },
  "2": Object {
    "createdAt": "2021-03-04T19:53:53.946Z",
    "description": "needs-based intangible",
    "id": "2",
    "name": "Garthburgh",
    "userCount": 94343,
    "views": 0,
  },
  "20": Object {
    "createdAt": "2021-03-05T02:11:23.683Z",
    "description": "Specialist",
    "id": "20",
    "name": "Burniceside",
    "userCount": 21446,
    "views": 0,
  },
  "3": Object {
    "createdAt": "2021-03-04T17:08:15.954Z",
    "description": "hub Configuration",
    "id": "3",
    "name": "New Raina",
    "userCount": 66367,
    "views": 0,
  },
  "4": Object {
    "createdAt": "2021-03-04T18:29:55.591Z",
    "description": "Creative",
    "id": "4",
    "name": "Fadelhaven",
    "userCount": 32882,
    "views": 0,
  },
  "5": Object {
    "createdAt": "2021-03-05T06:45:19.441Z",
    "description": "Coordinator invoice HTTP",
    "id": "5",
    "name": "Rachaelville",
    "userCount": 53290,
    "views": 0,
  },
  "6": Object {
    "createdAt": "2021-03-05T07:09:57.238Z",
    "description": "Computers",
    "id": "6",
    "name": "West Evashire",
    "userCount": 74569,
    "views": 0,
  },
  "7": Object {
    "createdAt": "2021-03-05T05:30:41.235Z",
    "description": "RSS solid state Refined Steel Table",
    "id": "7",
    "name": "Imeldamouth",
    "userCount": 42623,
    "views": 0,
  },
  "8": Object {
    "createdAt": "2021-03-04T15:25:20.453Z",
    "description": "matrix Heard Island and McDonald Islands",
    "id": "8",
    "name": "New Dollybury",
    "userCount": 31953,
    "views": 0,
  },
  "9": Object {
    "createdAt": "2021-03-05T07:26:18.807Z",
    "description": "Granite",
    "id": "9",
    "name": "Adelefurt",
    "userCount": 36318,
    "views": 0,
  },
}
`;
